### ISSUE TITLE
Implement Send in Lua

### ISSUE THREAD

**Issue by Lucaskyy** [khvzak/mlua#69]
Hi there.

When a Rust function is called from Lua (I am talking about the Lua argument you get from function calls) I want to use `tokio::spawn` to spawn a new thread and in there use the Lua state to get argument data. I am not able to do this currently because Lua cannot be moved safely because Send isn't implemented.

Is it possible to implement Send into Lua?

**Comment by Lucaskyy** [comment]
Actually, does this already exist? I saw a "send" feature flag.

**Comment by khvzak** [comment]
Could you provide a code example please?
In most cases you could use `tokio::task::spawn_local` to spawn `!Send` features.

Lua can be `send` (via the corresponding feature flag),  but not `Sync` unfortunately (yet?).

**Comment by Lucaskyy** [comment]
Hi, I didn't know the send feature existed :P I found it and closed this issue. I am currently however not at my PC. But Send is good enough for spawn, I just had Sync in my head (bad memory). I'll send you an example when I'm at my PC tomorrow.

**Comment by Lucaskyy** [comment]
Okay, I added the send feature flag:
```toml
[dependencies]
mlua = { version = "0.6", features = ["lua54", "module", "send"] }
```
This is my code:
```rust
fn send_request(lua: &Lua, opts: LuaTable) -> LuaResult<()> {
    let mut cb: LuaFunction = lua.create_function(|_, _: String| { Ok(()) })?;
    if opts.contains_key("callback").unwrap() {
        cb = opts.get::<_, LuaFunction>("callback").unwrap();
    }

    tokio::spawn(async move {
        // do something
        let s = res.text().await.unwrap();
        cb.call(s);
    });

    Ok(())
}
```
I then get this error:
```rust
error: future cannot be sent between threads safely
   --> src\lib.rs:30:5
    |
30  |     tokio::spawn(async move {
    |     ^^^^^^^^^^^^ future created by async block is not `Send`
    |
   ::: C:\Users\<USER>\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.10.0\src\task\spawn.rs:127:21
    |
127 |         T: Future + Send + 'static,
    |                     ---- required by this bound in `tokio::spawn`
    |
    = help: within `mlua::Lua`, the trait `Sync` is not implemented for `*mut c_void`
note: captured value is not `Send`
   --> src\lib.rs:34:9
    |
34  |         cb.call(s);
    |         ^^ has type `mlua::Function<'_>` which is not `Send`

error: future cannot be sent between threads safely
   --> src\lib.rs:30:5
    |
30  |     tokio::spawn(async move {
    |     ^^^^^^^^^^^^ future created by async block is not `Send`
    |
   ::: C:\Users\<USER>\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.10.0\src\task\spawn.rs:127:21
    |
127 |         T: Future + Send + 'static,
    |                     ---- required by this bound in `tokio::spawn`
    |
    = help: within `mlua::Lua`, the trait `Sync` is not implemented for `*mut mlua::lua::ExtraData`
note: captured value is not `Send`
   --> src\lib.rs:34:9
    |
34  |         cb.call(s);
    |         ^^ has type `mlua::Function<'_>` which is not `Send`

error: future cannot be sent between threads safely
   --> src\lib.rs:30:5
    |
30  |     tokio::spawn(async move {
    |     ^^^^^^^^^^^^ future created by async block is not `Send`
    |
   ::: C:\Users\<USER>\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.10.0\src\task\spawn.rs:127:21
    |
127 |         T: Future + Send + 'static,
    |                     ---- required by this bound in `tokio::spawn`
    |
    = help: within `mlua::Lua`, the trait `Sync` is not implemented for `UnsafeCell<()>`
note: captured value is not `Send`
   --> src\lib.rs:34:9
    |
34  |         cb.call(s);
    |         ^^ has type `mlua::Function<'_>` which is not `Send`
```

**Comment by Lucaskyy** [comment]
Using `tokio::task::spawn_local` fixes the issue above, but then gives this error:
```rust
error[E0621]: explicit lifetime required in the type of `lua`
  --> src\lib.rs:30:5
   |
30 |     tokio::task::spawn_local(async move {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ lifetime `'static` required
```

**Comment by Lucaskyy** [comment]
Also, just for reference. I am making a module and when my function is called, I want to set some variables up which I will be using. Then I want to use Tokio to spawn a new task/thread and run the code in there. Then, I want to use my `cb` variable which is a `Function` to call the callback specified in the `Table` argument.

**Comment by khvzak** [comment]
You still need to run a reactor (tokio runtime) and block somewhere to drive futures execution (even joining spawned background tasks).

An example how to use tokio in a module mode (without callbacks):
```rust
use mlua::{ExternalResult, Function, Lua, Result};

#[mlua::lua_module]
fn my_module(lua: &Lua) -> Result<Function> {
    lua.create_function(|lua, cb: Function| {
        let rt = tokio::runtime::Runtime::new().to_lua_err()?;

        let get = lua.create_async_function(|_, uri: String| async move {
            reqwest::get(&uri)
                .await
                .and_then(|r| r.error_for_status())
                .to_lua_err()?
                .text()
                .await
                .to_lua_err()
        })?;

        rt.block_on(async move { cb.call_async::<_, ()>(get).await })
    })
}
```

And example usage:
```
$ lua5.4 -l my_module -e 'my_module(function (get) print(get("https://httpbin.org/get")) end)'
```

**Comment by khvzak** [comment]
An example how to run multilple tasks in parallel using the modified code above:
```rust
        rt.block_on(async move {
            let tasks = cb.call_async::<_, Table>(get).await?;
            let mut futs = Vec::new();
            for task in tasks.sequence_values::<Function>() {
                futs.push(task?.call_async::<_, ()>(()));
            }
            futures::future::try_join_all(futs).await?;
            Ok(())
        })
```

Usage:
```lua
local f = require("my_module")

local tokio = require("my_module")

tokio(function (get)
    return {
        function() print(get("https://httpbin.org/get?a=b")) end,
        function() print(get("https://httpbin.org/get?b=a")) end,
    }
end)
```

**Comment by Lucaskyy** [comment]
Ahhh, thank you so much! 😄

**Comment by Lucaskyy** [comment]
Is it possible to run a task async, and invoke the callback later?
So it immediately returns, but then calls the callback after some time.

**Comment by khvzak** [comment]
You can do this (but still must run everything inside an event loop).

Rust:
```rust
use mlua::{ExternalResult, Function, Lua, Result};
use std::sync::{Arc, Mutex};

#[mlua::lua_module]
fn my_module(lua: &Lua) -> Result<Function> {
    lua.create_function(|lua, func: Function| {
        let rt = tokio::runtime::Runtime::new().to_lua_err()?;
        rt.block_on(async {
            let lua: &'static Lua = unsafe { std::mem::transmute(lua) };
            let tasks = Arc::new(Mutex::new(Vec::new()));

            let tasks2 = tasks.clone();
            let get = lua.create_function(move |_, (uri, cb): (String, Function)| {
                let mut tasks = tasks2.lock().unwrap();
                tasks.push(tokio::task::spawn_local(async move {
                    let res = reqwest::get(&uri).await.and_then(|r| r.error_for_status());
                    let body = res.to_lua_err()?.text().await.to_lua_err()?;
                    cb.call::<_, ()>(body)
                }));
                Ok(())
            })?;

            let local = tokio::task::LocalSet::new();
            local
                .run_until(async move {
                    func.call(get)?;
                    futures::future::join_all(&mut *tasks.lock().unwrap()).await;
                    Ok(())
                })
                .await
        })
    })
}
```

and Lua:
```lua
local tokio = require("my_module")

tokio(function (get)
    get("https://httpbin.org/get?a=b", function (body)
        print(body)
    end)
    get("https://httpbin.org/get?b=a", function (body)
        print(body)
    end)
end)
```

**Comment by Lucaskyy** [comment]
Thank you! This is exactly what I wanted :) Thanks for helping me and making this awesome library 😄
